lib/bb/taskdata.py

```diff
@@ -0,0 +1,586 @@
+        """
+        Maintain a list of build targets
+        """
+        if name not in self.build_targets:
+            self.build_targets[name] = []
+
+    def add_build_target(self, fn, item):
+        """
+        Add a build target.
+        If already present, append the provider fn to the list
+        """
+        if item in self.build_targets:
+            if fn in self.build_targets[item]:
+                return
+            self.build_targets[item].append(fn)
+            return
+        self.build_targets[item] = [fn]
+
+    def seen_run_target(self, name):
+        """
+        Maintain a list of runtime build targets
+        """
+        if name not in self.run_targets:
+            self.run_targets[name] = []
+
+    def add_runtime_target(self, fn, item):
+        """
+        Add a runtime target.
+        If already present, append the provider fn to the list
+        """
+        if item in self.run_targets:
+            if fn in self.run_targets[item]:
+                return
+            self.run_targets[item].append(fn)
+            return
+        self.run_targets[item] = [fn]
+
+    def mark_external_target(self, target):
+        """
+        Mark a build target as being externally requested
+        """
+        if target not in self.external_targets:
+            self.external_targets.append(target)
+
+    def get_unresolved_build_targets(self, dataCache):
+        """
+        Return a list of build targets who's providers
+        are unknown.
+        """
+        unresolved = []
+        for target in self.build_targets:
+            if re_match_strings(target, dataCache.ignored_dependencies):
+                continue
+            if target in self.failed_deps:
+                continue
+            if not self.build_targets[target]:
+                unresolved.append(target)
+        return unresolved
+
+    def get_unresolved_run_targets(self, dataCache):
+        """
+        Return a list of runtime targets who's providers
+        are unknown.
+        """
+        unresolved = []
+        for target in self.run_targets:
+            if re_match_strings(target, dataCache.ignored_dependencies):
+                continue
+            if target in self.failed_rdeps:
+                continue
+            if not self.run_targets[target]:
+                unresolved.append(target)
+        return unresolved
+
+    def get_provider(self, item):
+        """
+        Return a list of providers of item
+        """
+        return self.build_targets[item]
+
+    def get_dependees(self, item):
+        """
+        Return a list of targets which depend on item
+        """
+        dependees = []
+        for fn in self.depids:
+            if item in self.depids[fn]:
+                dependees.append(fn)
+        return dependees
+
+    def get_rdependees(self, item):
+        """
+        Return a list of targets which depend on runtime item
+        """
+        dependees = []
+        for fn in self.rdepids:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)